rs/index/src/posting_list/file.rs

```diff
@@ -0,0 +1,673 @@
+            let bytes_left_in_mmap = mmap.len() - current_offset;
+            let elems_in_mmap = std::cmp::min(remaining_elem, bytes_left_in_mmap / usize_in_bytes);
+
+            let slice = &mmap[current_offset..current_offset + elems_in_mmap * usize_in_bytes];
+            posting_list.extend_from_slice(unsafe {
+                std::slice::from_raw_parts(slice.as_ptr() as *const usize, elems_in_mmap)
+            });
+
+            remaining_elem -= elems_in_mmap;
+
+            if remaining_elem > 0 {
+                current_file_num += 1;
+                current_offset = 0;
+                if current_file_num >= self.mmaps.len() {
+                    return Err(anyhow!("Current file nunber out of bound"));
+                }
+            }
+        }
+
+        Ok(posting_list)
+    }
+
+    fn append(&mut self, posting_list: &[usize]) -> Result<()> {
+        if self.current_num_of_posting_list == self.num_clusters {
+            return Err(anyhow!(
+                "Trying to append more posting lists than number of clusters"
+            ));
+        }
+        let required_size = posting_list.len() * std::mem::size_of::<usize>();
+        self.size_bytes += required_size;
+        let should_flush = self.resident && self.size_bytes > self.memory_threshold;
+        let flush = should_flush && !self.resident_posting_lists.is_empty();
+
+        // Good case, where file is still resident
+        if self.resident && !should_flush {
+            self.current_num_of_posting_list += 1;
+            self.resident_posting_lists.push(posting_list.to_vec());
+            return Ok(());
+        }
+
+        // Spill to disk, creating new files if necessary
+        let pl_metadata_in_bytes = 2 * std::mem::size_of::<usize>();
+        if flush {
+            self.flush_resident_posting_lists_to_disk()?;
+            // At this point we are not in memory anymore, we'll need to
+            // take into account the storage for metadata.
+            self.size_bytes += self.resident_posting_lists.len() * pl_metadata_in_bytes;
+        }
+
+        self.size_bytes += pl_metadata_in_bytes;
+        // We should spill to disk, but did not flush (there was nothing to flush)
+        if self.resident && !flush {
+            // Trigger the creation of a new file since this is the first time we write
+            // to disk
+            self.current_offset = self.backing_file_size;
+            // Flip the flag since we are technically on disk at this point
+            self.resident = false;
+        } else {
+            // Adjust the offset to write metadata first
+            self.current_offset = self.current_num_of_posting_list * pl_metadata_in_bytes;
+        }
+        self.append_posting_list_to_disk(posting_list)?;
+        Ok(())
+    }
+
+    fn write(&mut self, writer: &mut std::io::BufWriter<&mut std::fs::File>) -> Result<usize> {
+        writer.write_all(&self.len().to_le_bytes())?;
```

Comment:
use `wrap_write` in `utils`, which will return the length. do it for all occurrences of write